File: functions/exchange_rate/domain/models/exchange_rate.py
from datetime import datetime

from exchange_rate.domain.value_objects import ExchangeID, Currency, Rate
from uuid6 import uuid7


class ExchangeRate:
    def __init__(
        self,
        exchange_id: ExchangeID | None,
        currency: Currency,
        rate: Rate | None,
        registered_at: datetime,
    ):
        self.exchange_id: ExchangeID | None = exchange_id
        self.currency: Currency = currency
        self.rate: Rate | None = rate
        self.registered_at: datetime = registered_at

    @staticmethod
    def create(
        currency: Currency,
        rate: Rate | None,
        registered_at: datetime,
    ):
        exchange_id = ExchangeID(uuid7())
        return ExchangeRate(exchange_id, currency, rate, registered_at)

File: functions-ts/src/exchange-rate/domain/models/exchange-rate.ts
import { v7 as uuid7 } from 'uuid'
import { z } from 'zod'

export class ExchangeRate {
  readonly id: string
  readonly currency: Currency
  readonly rate: number | null
  readonly registered_at: Date

  private constructor(
    id: string,
    currency: string,
    rate: number | null,
    registered_at: Date,
  ) {
    this.id = z.string().nonempty().parse(id)
    this.currency = z.enum(CurrencyValues).parse(currency)
    this.rate = rate
    this.registered_at = registered_at
  }

  toJSON() {
    return {
      id: this.id,
      currency: this.currency,
      rate: this.rate,
      registered_at: this.registered_at.toJSON(),
    }
  }

  static fromJSON(
    id: string,
    currency: string,
    rate: number,
    registered_at: string,
  ) {
    return new ExchangeRate(id, currency, rate, new Date(registered_at))
  }

  static create(currency: string, rate: number | null, registered_at: Date) {
    const id = uuid7()
    return new ExchangeRate(id, currency, rate, registered_at)
  }
}

export const CurrencyValues = ['USD', 'EUR', 'RUB', 'TRY', 'CNY'] as const

export type Currency = (typeof CurrencyValues)[number]
